### DISCUSSION TITLE
Working Group - Enhanced TLS support

### DISCUSSION THREAD

**Discussion by cescoffier** [quarkusio/quarkus#41024] [Design Discussions]
Hello,

We recently merged an ancient issue (3y old) to centralize the TLS configuration. Extensions should now be able to use this centralized registry to configure TLS. It avoids heterogeneous and incomplete configuration. 

This was a significant effort to simplify and improve TLS management. Typically, with this registry, we can homogenize the TLS configuration, add support for certificate reloading, integrate certificate providers like Let's Encrypt or Cert-Manager, and improve the local experience.

This effort is now tracked using a "Focus Group" (temporary wording) project: https://github.com/orgs/quarkusio/projects/24. 

I'm creating this discussion to raise awareness and as a place to discuss ideas, challenges, progress...

**Comment by emattheis** [comment]
@cescoffier I have some specialized requirements for setting up TLS using a private CA and I'm looking for a way to configure and periodically reload certificates and keys for the vert.x https server managed by Quarkus. I found [this post](https://github.com/quarkusio/quarkus/discussions/33142) showing how to set things up at configuration time and it works well, but there's no obvious way to introduce new certificates as Quarkus doesn't expose the resulting `HttpServer`.

I came up with a hacky solution using a custom KeyCertOptions implementation in conjunction with the built-in `quarkus.http.ssl.certificate.reload-period` functionality.

```java
import io.netty.handler.ssl.util.SelfSignedCertificate;
import io.quarkus.vertx.http.HttpServerOptionsCustomizer;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.core.net.KeyCertOptions;
import io.vertx.core.net.PemKeyCertOptions;
import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.net.ssl.KeyManagerFactory;
import java.security.cert.CertificateException;
import java.time.Duration;
import java.time.Instant;
import java.util.Date;
import java.util.function.Function;

@ApplicationScoped
public class CertificateManager implements HttpServerOptionsCustomizer {

    private static final Logger log = LoggerFactory.getLogger(CertificateManager.class);

    private final Duration certificateReloadPeriod;

    private volatile PemKeyCertOptions keyCertOptions = null;
    private volatile Instant keyCertOptionsExpiration = Instant.now();

    public CertificateManager(@ConfigProperty(name = "quarkus.http.ssl.certificate.reload-period") Duration certificateReloadPeriod) {
        this.certificateReloadPeriod = certificateReloadPeriod;
    }

    @Override
    public void customizeHttpsServer(HttpServerOptions options) {
        options.setKeyCertOptions(keyCertOptions());
    }

    private void generateCertificatesIfNecessary() {
        var nextReloadTime = Instant.now().plus(certificateReloadPeriod);
        if (keyCertOptionsExpiration.isBefore(nextReloadTime)) {
            var keyCertOptions = new PemKeyCertOptions();
            var padding = certificateReloadPeriod.dividedBy(2);
            var keyCertOptionsExpiration = Instant.now().plus(certificateReloadPeriod).plus(padding);
            var host = "localhost";
            var ecBits = 256;
            var rsaBits = 2048;
            var notBefore = Date.from(Instant.now().minus(padding));
            var notAfter = Date.from(keyCertOptionsExpiration);
            try {
                var ecCert = new SelfSignedCertificate(host, notBefore, notAfter, "EC", ecBits);
                var rsaCert = new SelfSignedCertificate(host, notBefore, notAfter, "RSA", rsaBits);
                log.info("generated self-signed certificates");
                keyCertOptions
                        .addCertPath(ecCert.certificate().getAbsolutePath())
                        .addKeyPath(ecCert.privateKey().getAbsolutePath())
                        .addCertPath(rsaCert.certificate().getAbsolutePath())
                        .addKeyPath(rsaCert.privateKey().getAbsolutePath());
            } catch (CertificateException e) {
                log.error("error generating self-signed certificates", e);
            }
            this.keyCertOptions = keyCertOptions;
            this.keyCertOptionsExpiration = keyCertOptionsExpiration;
        }
    }

    private KeyCertOptions keyCertOptions() {
        return new KeyCertOptions() {
            @Override
            public KeyCertOptions copy() {
                return keyCertOptions();
            }

            @Override
            public KeyManagerFactory getKeyManagerFactory(Vertx vertx) throws Exception {
                generateCertificatesIfNecessary();
                return keyCertOptions.getKeyManagerFactory(vertx);
            }

            @Override
            public Function<String, KeyManagerFactory> keyManagerFactoryMapper(Vertx vertx) throws Exception {
                generateCertificatesIfNecessary();
                return keyCertOptions.keyManagerFactoryMapper(vertx);
            }
        };
    }
}
```

It would be nice to be able to use the tls registry for this and take advantage of the in-built reload mechanism, but I can't see a way to register a `TlsConfiguration` programmatically that can be used by the built-in `HttpServer`. Am I missing something?

**Reply by cescoffier** [reply]
Hello,

The registry has a programmatic approach to register a TLS configuration (`io.quarkus.tls.TlsConfigurationRegistry#register`). You need to configure the HTTP server to use the given name. 

You may have to provide your own TLS configuration implementation. From what I see, you should extend `io.quarkus.tls.runtime.VertxCertificateHolder` - especially the `reload` method. 

While I am looking at the code, I think we could make this slightly simpler.

**Reply by emattheis** [reply]
@cescoffier yes, the trouble is the HTTP server validates that the config exists and application code doesn’t get a chance to register it.

I’ve been digging into the TLS-config extension a little and I think it would be interesting to allow beans to provide key/trust-stores.

**Reply by cescoffier** [reply]
You will need to register it before the start of the HTTP server, as you did with the customizer. 

I agree that we could reverse the injection using a whiteboard pattern (plugging any bean exposing a given interface). If you want to contribute it, I can help.

**Reply by emattheis** [reply]
> You will need to register it before the start of the HTTP server, as you did with the customizer.

This does not work, unfortunately, because the runtime [validates that the config exists](https://github.com/quarkusio/quarkus/blob/ff6047fe8ac0b3022a861682ce2024cd2c84ed74/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L800-L820) _before_ invoking the customizer.

> I agree that we could reverse the injection using a whiteboard pattern (plugging any bean exposing a given interface). If you want to contribute it, I can help.

I've been working on a prototype while familiarizing myself with the tls-registry internals. One thing that stands out right now is the strange relationship between `KeyStore`s and `KeyCertOptions`/`TrustOptions`. The existing implementations all start from Vert.x implementations and then expose `KeyStore`s from them. This sets up an assumption that all `TrustOptions` will be `KeyStore` based, which is highly likely of course, but makes the interface a bit awkward.

It's not clear to me when configurations are expected to provide `KeyStore`s or not, but the obvious thing to allow pluggable `KeyCertOption`s and `TrustOptions` would be to let beans provide `KeyManagerFactory` and `TrustManagerFactory` then use the corresponding `wrap` methods in the Vert.x interfaces.

**Reply by cescoffier** [reply]
Sorry, you need a `VertxOptionsCustomizer`. 

About the second point, when you use javax.ssl option, you provide a keystore directly.

**Reply by emattheis** [reply]
> Sorry, you need a `VertxOptionsCustomizer`.

Hmmm. That doesn't seem to work, either. I'm getting this at startup:

```
jakarta.enterprise.inject.CreationException: Synthetic bean instance for io.quarkus.tls.TlsConfigurationRegistry not initialized yet: io_quarkus_tls_TlsConfigurationRegistry_c3690e7f4866a79befda78c4c24a5b11fe6d5e3a
        - a synthetic bean initialized during RUNTIME_INIT must not be accessed during STATIC_INIT
        - RUNTIME_INIT build steps that require access to synthetic beans initialized during RUNTIME_INIT should consume the SyntheticBeansRuntimeInitBuildItem
```

I tried this:

```java
import io.netty.handler.ssl.util.SelfSignedCertificate;
import io.quarkus.tls.BaseTlsConfiguration;
import io.quarkus.tls.TlsConfigurationRegistry;
import io.quarkus.vertx.VertxOptionsCustomizer;
import io.vertx.core.VertxOptions;
import io.vertx.core.net.KeyCertOptions;
import io.vertx.core.net.PemKeyCertOptions;
import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.security.cert.CertificateException;
import java.util.Optional;

@ApplicationScoped
public class CertificateManager implements VertxOptionsCustomizer {

    private static final Logger log = LoggerFactory.getLogger(CertificateManager.class);

    private final TlsConfigurationRegistry tlsConfigurationRegistry;
    private final String httpTslConfigurationName;

    public CertificateManager(TlsConfigurationRegistry tlsConfigurationRegistry,
                              @ConfigProperty(name = "quarkus.http.tls-configuration-name") Optional<String> httpTslConfigurationName) {
        this.tlsConfigurationRegistry = tlsConfigurationRegistry;
        this.httpTslConfigurationName = httpTslConfigurationName.orElse(null);
    }

    @Override
    public void accept(VertxOptions vertxOptions) {
        if (httpTslConfigurationName != null && tlsConfigurationRegistry.get(httpTslConfigurationName).isEmpty()) {
            tlsConfigurationRegistry.register(httpTslConfigurationName, new BaseTlsConfiguration() {
                @Override
                public KeyCertOptions getKeyStoreOptions() {
                    var keyCertOptions = new PemKeyCertOptions();
                    var host = "localhost";
                    var ecBits = 256;
                    var rsaBits = 2048;
                    try {
                        var ecCert = new SelfSignedCertificate(host, "EC", ecBits);
                        var rsaCert = new SelfSignedCertificate(host, "RSA", rsaBits);
                        log.info("generated self-signed certificates");
                        keyCertOptions
                                .addCertPath(ecCert.certificate().getAbsolutePath())
                                .addKeyPath(ecCert.privateKey().getAbsolutePath())
                                .addCertPath(rsaCert.certificate().getAbsolutePath())
                                .addKeyPath(rsaCert.privateKey().getAbsolutePath());
                        return keyCertOptions;
                    } catch (CertificateException e) {
                        throw new RuntimeException("exception generating self-signed certificates", e);
                    }
                }
            });
        }
    }
}
```

**Reply by cescoffier** [reply]
Oh right, the TLS registry is not yet initialized. We do not _fire_ the Vert.x instance as a CDI event (which could have been helpful here). 

@mkouba do you know if we can be notified between the Vert.x initialization and the HTTP server startup?

**Reply by emattheis** [reply]
~HTTP server definitely fires events, but by then it's too late 😉.~

[edit]: Never mind, I see you said _between_ the Vert.x init and HTTP server startup.

Producing the TLS configuration directly is not ideal anyway. I really just want to supply `KeyStore`s from code and by able to configure the rest of the behavior in config. I think maybe we can just support something like this:

```
quarkus.tls.key-store.provider.bean-name
quarkus.tls.trust-store.provider.bean-name
```

Then we lookup the bean at run time and use it to get a KeyStore on demand, which can be wrapped into a corresponding `KeyCertOptions` or `TrustOptions` as necessary.

**Reply by cescoffier** [reply]
Yes, that would work, and it's relatively easy to add. Instead of bean-name, I would use `identifier` and use the `@Identifier` qualifier. Otherwise, `classname`.

**Reply by mkouba** [reply]
> @mkouba do you know if we can be notified between the Vert.x initialization and the HTTP server startup?

Hm, `quarkus-vertx` does produce a `ServiceStartBuildItem` so the `StartupEvent` should be always fired after the Vertx instance is initialized but I'm not 100% sure we can guarantee that it will be fired before the HTTP server starts...

**Reply by emattheis** [reply]
@cescoffier providing `KeyStore` directly turns out to be a bit of a pain. To align well with the existing internals, I ended up introducing these interfaces:

```java
package io.quarkus.tls;

import io.quarkus.tls.runtime.KeyStoreAndKeyCertOptions;
import io.vertx.core.Vertx;

public interface KeyStoreProvider {

    KeyStoreAndKeyCertOptions getKeyStore(Vertx vertx);
}
```

```java
package io.quarkus.tls;

import io.quarkus.tls.runtime.TrustStoreAndTrustOptions;
import io.vertx.core.Vertx;

public interface TrustStoreProvider {

    TrustStoreAndTrustOptions getTrustStore(Vertx vertx);
}
```

This gives the developer the freedom to decide exactly what to expose, while preserving the normal behavior internally, and I can now do this:

```java
import io.netty.handler.ssl.util.SelfSignedCertificate;
import io.quarkus.tls.KeyStoreProvider;
import io.quarkus.tls.runtime.KeyStoreAndKeyCertOptions;
import io.vertx.core.Vertx;
import io.vertx.core.net.PemKeyCertOptions;
import jakarta.enterprise.context.ApplicationScoped;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@ApplicationScoped
public class CertificateManager implements KeyStoreProvider {

    private static final Logger log = LoggerFactory.getLogger(CertificateManager.class);

    @Override
    public KeyStoreAndKeyCertOptions getKeyStore(Vertx vertx) {
        var keyCertOptions = new PemKeyCertOptions();
        var host = "localhost";
        var ecBits = 256;
        var rsaBits = 2048;
        try {
            var ecCert = new SelfSignedCertificate(host, "EC", ecBits);
            var rsaCert = new SelfSignedCertificate(host, "RSA", rsaBits);
            log.info("generated self-signed certificates");
            keyCertOptions
                    .addCertPath(ecCert.certificate().getAbsolutePath())
                    .addKeyPath(ecCert.privateKey().getAbsolutePath())
                    .addCertPath(rsaCert.certificate().getAbsolutePath())
                    .addKeyPath(rsaCert.privateKey().getAbsolutePath());
            return new KeyStoreAndKeyCertOptions(keyCertOptions.loadKeyStore(vertx), keyCertOptions);
        } catch (Exception e) {
            throw new RuntimeException("exception generating self-signed certificates", e);
        }
    }
}
```

Under the covers I changed the `CertificateRecorder` to look for beans if trust/key stores are not already configured. As you suggested, the `@Identifier` annotation is used to bind the provider to a named TLS config, and omitting it will match the default config.

It also scans for `@Identifier` qualifiers on provider beans at startup to automatically register `TlsConfiguraton` instances that aren't explicitly mentioned in the config.

I'll push the branch to my fork today and I'll be happy to submit a PR if you think it makes sense.

**Reply by emattheis** [reply]
Here are the changes: https://github.com/emattheis/quarkus/commit/7465ebbb9c200f8f201752403f6c6119079f3a2b

**Reply by cescoffier** [reply]
Awesome! Can you open a PR? It will make reviewing smoother!

**Reply by emattheis** [reply]
My pleasure! #45937

**Reply by emattheis** [reply]
@cescoffier Thanks for the feedback. I didn't want to go too deep without a first review, but now that I know it's acceptable I'll work on some documentation.

**Reply by cescoffier** [reply]
Awesome, thank you so much. It's a great contribution.
